metrics/serve.go

```diff
@@ -103,19 +103,112 @@ func RunMetrics(metricCount, labelCount, seriesCount, metricLength, labelLength,
+	changeSeriesTick := time.NewTicker(time.Duration(seriesChangeInterval) * time.Second)
 	updateNotify := make(chan struct{}, 1)
 
+	var currentSeriesCount int
+
+	switch seriesOperationMode {
+	case "double-halve":
+		currentSeriesCount = seriesCount
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+		seriesIncrease := true
+		go func() {
+			for tick := range changeSeriesTick.C {
+				metricsMux.Lock()
+				unregisterMetrics()
+				registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+				cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+				metricsMux.Unlock()
+				if seriesIncrease {
+					currentSeriesCount *= 2
+				} else {
+					currentSeriesCount /= 2
+					if currentSeriesCount < 1 {
+						currentSeriesCount = 1
+					}
+				}
+
+				fmt.Printf("%v: Adjusting series count. New count: %d\n", tick, currentSeriesCount)
+
+				seriesIncrease = !seriesIncrease
+
+				select {
+				case updateNotify <- struct{}{}:
+				default:
+				}
+			}
+		}()
+
+	case "gradual-change":
+		if minSeriesCount >= maxSeriesCount {
+			return nil, fmt.Errorf("error: minSeriesCount must be less than maxSeriesCount, got %d and %d", minSeriesCount, maxSeriesCount)
+		}
+		seriesIncrease := true
+		currentSeriesCount = minSeriesCount
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+		go func() {
+			for tick := range changeSeriesTick.C {
+				metricsMux.Lock()
+				unregisterMetrics()
+				registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+				cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+				metricsMux.Unlock()
+				if seriesIncrease {
+					currentSeriesCount += seriesChangeRate
+					if currentSeriesCount >= maxSeriesCount {
+						currentSeriesCount = maxSeriesCount
+						seriesIncrease = false
+					}
+				} else {
+					currentSeriesCount -= seriesChangeRate
+					if currentSeriesCount <= minSeriesCount {
+						currentSeriesCount = minSeriesCount
+						seriesIncrease = true
+					}
+				}
+				fmt.Printf("%v: Adjusting series count. New count: %d\n", tick, currentSeriesCount)
+
+				select {
+				case updateNotify <- struct{}{}:
+				default:
+				}
+			}
+		}()
+	default:
+		currentSeriesCount = seriesCount
+		registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+		cycleValues(labelKeys, labelValues, seriesCount, seriesCycle)
+		go func() {
+			for tick := range metricTick.C {
+				metricsMux.Lock()
+				fmt.Printf("%v: refreshing metric cycle\n", tick)
+				metricCycle++
+				unregisterMetrics()
+				registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+				cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+				metricsMux.Unlock()
+
+				select {
+				case updateNotify <- struct{}{}:
+				default:
+				}
+			}
+		}()
+	}
+
 	go func() {
```

Comment:
we should think about whether to move the value tick and series ticks to their own functions as well, and call them inside each of the new operation mode cases in the switch statement _if_ they are needed